include/aos/iam/certhandler.hpp

```diff
@@ -8,12 +8,110 @@
 #ifndef AOS_CERTHANDLER_HPP_
 #define AOS_CERTHANDLER_HPP_
 
+#include "aos/common/tools/array.hpp"
 #include "aos/common/tools/error.hpp"
+#include "aos/common/tools/function.hpp"
+#include "aos/common/tools/log.hpp"
+#include "aos/common/tools/string.hpp"
+#include "aos/common/tools/thread.hpp"
+#include "aos/common/tools/utils.hpp"
+#include "aos/common/types.hpp"
+
+// Missing types
+// Time
+// config::ModuleConfig => from config.go
+// ASN1, x509
+// crypto::PrivateKey, crypto::Signer
+// Function<Ret, Args...>
+// aoscloud/aos_common/utils/cryptutils
 
 namespace aos {
 namespace iam {
 namespace certhandler {
 
+/***********************************************************************************************************************
+ *
+ **********************************************************************************************************************/
+struct Time {
+    static constexpr uint64_t Hour = 60 * 60 * 1000; // ms.
+};
+
+namespace crypto {
+using PrivateKey = String;
+}
+
+namespace x509 {
+using Certificate = Array<uint8_t>;
+}
+
+namespace config {
+struct ModuleConfig { };
+} // namespace config
+
+namespace json {
+using RawMessage = String;
+}
+
+template <class... Args>
+class TFunction { };
+
+// A new interface that provides some functionality required for a New/register/createModule methods in GO code.
+class ModuleItf { }; // Q: We already have CertModuleItf below, so may be  we could  integrate this functionality there.
+
+/***********************************************************************************************************************
+ * Types
+ **********************************************************************************************************************/
+
+using CertificateSerialNum = StaticString<AOS_CONFIG_TYPES_CERTIFICATE_SERIAL_LEN>;
+using CertificateType = StaticString<AOS_CONFIG_TYPES_CERTIFICATE_TYPE_LEN>;
+using CertificatePublicKey = StaticString<AOS_CONFIG_TYPES_CERTIFICATE_PUB_KEY_LEN>;
+using Issuer = StaticString<AOS_CONFIG_TYPES_CERTIFICATE_ISSUER_ID_LEN>;
+using Password = StaticString<AOS_CONFIG_TYPES_PASSWORD_LEN>;
+
+// CertInfo certificate info.
+struct CertInfo {
+    Issuer               mIssuer;
+    CertificateSerialNum mSerial;
+    Url                  mCertURL;
+    Url                  mKeyURL;
+    Time                 mNotAfter;
+};
+
+// CertStorage provides API to store/retrieve certificates info.
+class CertStorageItf {
+public:
+    virtual Error AddCertificate(const String& certType, const CertInfo& cert) = 0;
+    virtual Error GetCertificate(const String& issuer, const String& serial, CertInfo& res) = 0;
+    virtual Error GetCertificates(const String& certType, Array<CertInfo>& res) = 0;
+    virtual Error RemoveCertificate(const String& certType, const String& certURL) = 0;
+    virtual Error RemoveAllCertificates(const String& certType) = 0;
+    virtual ~CertStorageItf() = default;
+};
```

Comment:
In storage let's rename all methods as following:

`AddCertInfo`
`GetCertInfo`
etc.